sdk/clientcore/core/src/main/java/io/clientcore/core/implementation/http/rest/RestProxyImpl.java

```diff
@@ -243,7 +250,97 @@ public void updateRequest(RequestDataConfiguration requestDataConfiguration, Obj
                 request.setBody(BinaryData.fromBytes(array));
             }
         } else {
-            request.setBody(BinaryData.fromObject(bodyContentObject, serializerAdapter));
+            request.setBody(BinaryData.fromObject(bodyContentObject,
+                serializer.getSerializerForFormat(serializationFormatFromContentType(request.getHeaders()))));
         }
     }
+
+    /**
+     * Determines the serializer encoding to use based on the Content-Type header.
+     *
+     * @param headers the headers to get the Content-Type to check the encoding for.
+     * @return the serializer encoding to use for the body. {@link SerializationFormat#JSON} if there is no Content-Type
+     * header or an unrecognized Content-Type encoding is given.
+     */
+    public static SerializationFormat serializationFormatFromContentType(HttpHeaders headers) {
+        if (headers == null) {
+            return SerializationFormat.JSON;
+        }
+
+        String contentType = headers.getValue(HttpHeaderName.CONTENT_TYPE);
+        if (ImplUtils.isNullOrEmpty(contentType)) {
+            // When in doubt, JSON!
+            return SerializationFormat.JSON;
+        }
+
+        int contentTypeEnd = contentType.indexOf(';');
+        contentType = (contentTypeEnd == -1) ? contentType : contentType.substring(0, contentTypeEnd);
+        SerializationFormat encoding = checkForKnownEncoding(contentType);
+        if (encoding != null) {
+            return encoding;
+        }
+
+        int contentTypeTypeSplit = contentType.indexOf('/');
+        if (contentTypeTypeSplit == -1) {
+            return SerializationFormat.JSON;
+        }
+
+        // Check the suffix if it does not match the full types.
+        // Suffixes are defined by the Structured Syntax Suffix Registry
+        // https://www.rfc-editor.org/rfc/rfc6839
+        final String subtype = contentType.substring(contentTypeTypeSplit + 1);
+        final int lastIndex = subtype.lastIndexOf('+');
+        if (lastIndex == -1) {
+            return SerializationFormat.JSON;
+        }
+
+        // Only XML and JSON are supported suffixes, there is no suffix for TEXT.
+        final String mimeTypeSuffix = subtype.substring(lastIndex + 1);
+        if ("xml".equalsIgnoreCase(mimeTypeSuffix)) {
+            return SerializationFormat.XML;
+        } else if ("json".equalsIgnoreCase(mimeTypeSuffix)) {
+            return SerializationFormat.JSON;
```

Comment:
nit: Can we move this to a Util to make the class a bit leaner?

Reply:
I'm planning another PR in this area and will clean it up then. The `RestProxy` stuff is a bit complex being copied from `azure-core` where it has a sync and async variant. `clientcore` doesn't need to have `RestProxyBase`, `RestProxyImpl`, and `RestProxyUtil`, it can all be merged into `RestProxyImpl`.